nixos/modules/services/web-servers/garage.nix

```diff
@@ -0,0 +1,84 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.services.garage;
+  toml = pkgs.formats.toml {};
+  configFile = toml.generate "garage.toml" cfg.settings;
+in
+{
+  meta.maintainers = [ maintainers.raitobezarius ];
+
+  options.services.garage = {
+    enable = mkEnableOption (lib.mdDoc "Garage Object Storage (S3 compatible)");
+
+    logLevel = mkOption {
+      type = types.enum (["info" "debug" "trace"]);
+      default = "info";
+      example = "debug";
+      description = lib.mdDoc "Garage log level, see <https://garagehq.deuxfleurs.fr/documentation/quick-start/#launching-the-garage-server> for examples.";
+    };
+
+    settings = mkOption {
+      type = types.submodule {
+        freeformType = toml.type;
+
+        options = {
+          metadata_dir = mkOption {
+            default = "/var/lib/garage/meta";
+            type = types.path;
+            description = lib.mdDoc "The metadata directory, put this on a fast disk (e.g. SSD) if possible.";
+          };
+
+          data_dir = mkOption {
+            default = "/var/lib/garage/data";
+            type = types.path;
+            description = lib.mdDoc "The main data storage, put this on your large storage (e.g. high capacity HDD)";
+          };
+
+          replication_mode = mkOption {
+            default = "none";
+            type = types.enum ([ "none" "1" "2" "3" 1 2 3 ]);
+            apply = v: toString v;
+            description = lib.mdDoc "Garage replication mode, defaults to none, see: <https://garagehq.deuxfleurs.fr/reference_manual/configuration.html#replication_mode> for reference.";
+          };
+        };
+      };
+      description = lib.mdDoc "Garage configuration, see <https://garagehq.deuxfleurs.fr/reference_manual/configuration.html> for reference.";
+    };
+
+    package = mkOption {
+      default = pkgs.garage;
+      defaultText = literalExpression "pkgs.garage";
+      type = types.package;
+      description = lib.mdDoc "Garage package to use.";
+    };
+  };
+
+  config = mkIf cfg.enable {
+    environment.systemPackages = [ (pkgs.writeShellScriptBin "garage-manage" ''
+      export GARAGE_CONFIG_FILE="${configFile}"
+      exec ${cfg.package}/bin/garage "$@"
+    '') ]; # For administration
```

Comment:
To be fair, it makes multi-instance harder.
I prefer to scope it but I have no strong feeling.

Reply:
@SuperSandro2000 if I move this module to `services.garage.instances.xxx.`, I can have multi-instances by running N systemd services for Garage

Reply:
also, it may make easier to do things like major upgrades.

Reply:
No strong feeling on that, we can have `/etc/garage.toml` then revisit it later, what do you say @SuperSandro2000 ?

Reply:
Done

Reply:
> @SuperSandro2000 if I move this module to `services.garage.instances.xxx.`, I can have multi-instances by running N systemd services for Garage

Do we support something like that usually? I am not sure. Maybe we should just change the module to be multi instance by itself so that it is clear why things are done how they are done.

Reply:
Sure, let's do this in a future PR, @SuperSandro2000